### PR TITLE
Async query get result bug fix

### PR DESCRIPTION
**Pull request by dai-chen**
### Description

Followed option 2 in https://github.com/opensearch-project/sql/issues/2436#issuecomment-1834172587. Consider query statement complete and successful only if **statement state is SUCCESS and query result is available (searchable)**. Please see more details in the issue comment.

Changes:

1. Changed template workflow in abstract class `AsyncQueryHandler` and subclass as needed
2. Moved mock index class in `IndexQuerySpecTest` up to parent class so new IT in this PR can use it.
 
### Issues Resolved

https://github.com/opensearch-project/sql/issues/2436, https://github.com/opensearch-project/sql/issues/2439
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dai-chen** [comment]
The failed CI seems not relevant. It keeps failing on retry and failed in  `main` as well.

**Review by kaituo — APPROVED** [review]

File: spark/src/test/java/org/opensearch/sql/spark/asyncquery/AsyncQueryGetResultSpecTest.java [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    // Create mock index with index state refreshing
+    mockIndex.createIndex();
+    mockIndexState.refreshing();
+
+    LocalEMRSClient emrClient = new LocalEMRSClient();
+    emrClient.setJobState("Cancelled");
+    createAsyncQuery(mockIndex.query, emrClient)
+        .withInteraction(interaction -> new JSONObject()) // simulate result index refresh delay
+        .assertQueryResults("running", null)
+        .withoutInteraction()
```
Comment (penghuo) [inline spark/src/test/java/org/opensearch/sql/spark/asyncquery/AsyncQueryGetResultSpecTest.java:180]: withoutInteraction() naming is not clear. could we have better name?

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/IndexDMLHandler.java [hunk 2]
```diff
@@ -106,7 +109,11 @@ protected JSONObject getResponseFromResultIndex(AsyncQueryJobMetadata asyncQuery
 
   @Override
   protected JSONObject getResponseFromExecutor(AsyncQueryJobMetadata asyncQueryJobMetadata) {
-    throw new IllegalStateException("[BUG] can't fetch result of index DML query form server");
+    // Consider statement still running if result doc created in submit() is not available yet
+    JSONObject result = new JSONObject();
+    result.put(STATUS_FIELD, StatementState.RUNNING.getState());
+    result.put(ERROR_FIELD, "");
```
[the remark this answers is not in the document]
Comment (dai-chen) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/IndexDMLHandler.java:115]: In `getAsyncQueryResults()` API code, it always fetch status and error field from what is returned here. Although it uses `optString`, I thought it maybe more clear to pass it here.

**Review by YANG-DB — COMMENTED** [review]
see comments below

File: spark/src/test/java/org/opensearch/sql/spark/asyncquery/AsyncQueryGetResultSpecTest.java (hunk 1 above)
Comment (dai-chen) [reply spark/src/test/java/org/opensearch/sql/spark/asyncquery/AsyncQueryGetResultSpecTest.java]: Sure, removed and inlined the single-line for clarify in https://github.com/opensearch-project/sql/pull/2443/commits/0dc8b4d94ecaef6c5bb05259c85196359584440c. Thanks!

File: spark/src/test/java/org/opensearch/sql/spark/asyncquery/AsyncQueryExecutorServiceSpec.java [hunk 3]
```diff
@@ -215,6 +228,7 @@ public static class LocalEMRSClient implements EMRServerlessClient {
     private int startJobRunCalled = 0;
     private int cancelJobRunCalled = 0;
     private int getJobResult = 0;
+    private String jobState = "RUNNING";
```
[the remark this answers is not in the document]
Comment (dai-chen) [reply spark/src/test/java/org/opensearch/sql/spark/asyncquery/AsyncQueryExecutorServiceSpec.java]: Changed it to existing Enum `JobRunState` in https://github.com/opensearch-project/sql/pull/2443/commits/52499afeb7de50d3f85e4a589fdab4d823f49381. Thanks!

**Review by penghuo — APPROVED** [review]

**Review by vamsi-amazon — APPROVED** [review]

**Pull request merged**